### PR TITLE
Rewrite Query Language docs

### PR DESCRIPTION
**Pull request by ddelemeny**
Piggybacking on @trinity-1686a's update proposal for the query language documentation in #4708.

 I extracted and formatted a "friendly" version to be included in the _Getting started_ section of the docs and rewrote the refdoc in a more formal way

- Extract an introduction to the query language
- Rewrite the QL reference doc in a more systematic and exhaustive way.

##### Intro
image

##### Reference
image

### CODE REVIEW

**Comment by ddelemeny** [comment]
Okay, that should to the trick. Please make sure I didn't leave obvious mistakes in my interpretation.

@trinity-1686a I've left a few comments in the refdoc where I think some clarification is needed, feel free to chime in !

File: docs/get-started/query-language-intro.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+## Clauses
+
+The main concept of this language is a clause, which represents a simple condition that can be tested against documents. 
+
+### Querying fields
+
+A clause operates on fields of your document. It has the following syntax :
+```
+field: condition
```
Comment (trinity-1686a) [inline docs/get-started/query-language-intro.md:16]: there should be no space between `field:` and `condition`

**Review by trinity-1686a — COMMENTED** [review]
I don't think we have a name for `field:'my value'` queries. It's a bit like a phrase query in the sense that it matches only if field contains both my and value; but it doesn't check positions (i.e. they can be distant or not in the right order). It's also a convenient way to query fields with the raw tokenizer when the value contain spaces (which can also be done with `field:my\ value` I believe)

File: docs/reference/query-language.md [hunk 2]
```diff
@@ -1,108 +1,236 @@
...
+###### Limitation
 
-### Range queries
+Quickwit may trim some results matched by this clause in some cases.  If you search for `"thanks for your co"*`, it will enumerate the first 50 tokens which start with "co", and search for any documents where "thanks for your" is followed by any of these tokens.
 
-Range queries can only be executed on fields with a fast field. Currently only fields of type `ip` are supported.
+If there are many tokens starting with "co", "contribution" might not be one of the 50 selected tokens, and the query won't match a document containing "thanks for your contribution". Normal prefix queries don't suffer from this issue.
 
+
+<!-- NEEDS CLARIFICATION : what does "first 50 tokens" mean ? in what order ? can the value be tuned ? -->
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/reference/query-language.md [hunk 3]
```diff
@@ -1,108 +1,236 @@
...
-Quickwit uses a query mini-language which is used by providing a `query` parameter to the search endpoints.
+## Pseudo-grammar
+
+```
+query = '(' query ')'
+      | query operator query
+      | unary_operator query
+      | clause
+
+operator = 'AND' | 'OR'
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/reference/query-language.md [hunk 4]
```diff
@@ -1,108 +1,236 @@
...
+---
 
-For instance, the document `{"product": {"attributes": {color": "red"}}}` is returned if you query `product.attributes.color:red`.
+## Types of clauses
 
-If a dot `.` exists in one of the key of your object, the above syntax has some ambiguity.
-For instance, by default, `{"k8s.component.name": "quickwit"}` will be matched by `k8s.component.name:quickwit`.
+### Term `field:term`
+```
+term: term_char+
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/get-started/query-language-intro.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+| term prefix | `field: prefix*` | `app_name: tant*` <br/> `quick*` | A term clause tests the existence of a token starting with the provided value | yes |
+| term set | `field: IN [token token ..]` |`severity: IN [error warn]` | A term set clause tests the existence of any of the provided value in the field's tokens| yes |
+| phrase | `field: "sequence of tokens"` | `full_name: "john doe"` | A phrase clause tests the existence of the provided sequence of tokens | yes |
+| phrase prefix | `field: "sequence of tokens"*` | `title: "how to m"*` | A phrase prefix clause tests the exsitence of a sequence of tokens, the last one used like in a prefix clause | yes |
+| all | `*` | `*` | A match-all clause will match every document | no |
+| exist | `field: *` | `error: *` | An exist clause tests the existence of any value for the field, it will match only if the field exists | no |
+| range | `field: bounds` |`duration: [0 1000}` <br/> `last_name: [banner miller]` | A term clause tests the existence of a token between the provided bounds | no |
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/get-started/query-language-intro.md]: OK

File: docs/reference/query-language.md [hunk 6]
```diff
@@ -1,108 +1,236 @@
...
 
-Quickwit supports `AND`, `+`, `OR`, `NOT` and `-` as Boolean operators (case sensitive). By default, the `AND` is chosen, this means that if you omit it in a query like `title:"barack obama" president` Quickwit will interpret the query as `title:"barack obama" AND president`.
+Matches documents if the targeted field contains a token which starts with the provided value.
 
-### Grouping boolean operators
+`field:quick*` will match any document where the field 'field' has a token like `quickwit` or `quickstart`, but not `qui` or `abcd`.
 
-Quickwit supports parenthesis to group multiple clauses:
 
+### Term set `field: IN [a b c]`
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/reference/query-language.md [hunk 7]
```diff
@@ -1,108 +1,236 @@
...
+
+
+---
+
+## Other considerations 
+
+### Default Search Fields
+In many case it is possible to omit the field you search if it was configured in the `default_search_fields` array of the index configuration.
+
+<!-- NEED CLARIFICATION : default fields clauses behavior on an array is combined using OR or AND ? -->
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/reference/query-language.md [hunk 8]
```diff
@@ -1,108 +1,236 @@
...
+
+###### Inclusive and exclusive bounds
+Inclusive bounds are represented by square brackets `[]`. They will match tokens equal to the bound term.
+Exclusive bounds are represented by curly brackets `{}`. They will not match tokens equal to the bound term.
+
+###### Half-Open bounds
+You can make an half open range by using `*` as one of the bounds. `field:[b TO *]` will match 'bb' and 'zz', but not 'ab'.
+You can also use a comparison based syntax:`field:<b`, `field:>b`, `field:<=b` or `field:>=b`.
+
+<!-- NEEDS CLARIFICATION : ordering of empty values ? -->
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK. Side question: what lexicographic order are we talking about ? I'm assuming unicode ordinal values without transliteration.

File: docs/get-started/query-language-intro.md (hunk 1 above)
Comment (ddelemeny) [reply docs/get-started/query-language-intro.md]: OK

File: docs/reference/query-language.md [hunk 9]
```diff
@@ -1,108 +1,236 @@
...
 ```
-(color:red OR color:green) AND size:large
+term_set = 'IN' '[' term_list ']'
+term_list = term_list term
+          | term
 ```
+Matches if the document contains any of the tokens provided. 
+
+###### Examples
+`field: IN [ab cd]` will match 'ab' or 'cd', but nothing else.
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/reference/query-language.md [hunk 10]
```diff
@@ -1,108 +1,236 @@
...
+### Range `field: [low_bound high_bound}`
+```
+range = explicit_range | comparison_half_range
+
+explicit_range = left_bound_char bounds right_bound_char
+left_bound_char = '[' | '{' 
+right_bound_char = '}' | ']'
+bounds = term term
+       | term '*'
+       | '*' term
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/reference/query-language.md [hunk 11]
```diff
@@ -1,108 +1,236 @@
...
+Quickwit may trim some results matched by this clause in some cases.  If you search for `"thanks for your co"*`, it will enumerate the first 50 tokens which start with "co", and search for any documents where "thanks for your" is followed by any of these tokens.
 
-Range queries can only be executed on fields with a fast field. Currently only fields of type `ip` are supported.
+If there are many tokens starting with "co", "contribution" might not be one of the 50 selected tokens, and the query won't match a document containing "thanks for your contribution". Normal prefix queries don't suffer from this issue.
 
+
+<!-- NEEDS CLARIFICATION : what does "first 50 tokens" mean ? in what order ? can the value be tuned ? -->
+
+
+### Range `field: [low_bound high_bound}`
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/reference/query-language.md [hunk 12]
```diff
@@ -1,108 +1,236 @@
...
+```
+---
+## Writing Queries
+### Escaping Special Characters
 
-The `query` is parsed into a series of terms and operators. There are two types of terms: single terms such as “tantivy” and phrases which is a group of words surrounded by double quotes such as “hello world”.
+Special reserved characters are: `+` , `^`, `` ` ``, `:`, `{`, `}`, `"`, `[`, `]`, `(`, `)`, `~`, `!`, `\\`, `*`, `SPACE`. Such characters can still appear in query terms, but they need to be escaped by an anti-slash `\` .
 
-Multiple terms can be combined together with Boolean operators `AND, OR` to form a more complex query. By default, terms will be combined with the `AND` operator.
+<!-- NEED CLARIFICATION: where is escaping necessary ? non-quoted terms ? field names ?-->
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/reference/query-language.md [hunk 13]
```diff
@@ -1,108 +1,236 @@
...
 
-IP addresses can be provided as IpV4 or IpV6. It is recommended to use the same format as in the indexed documents.
+### Allowed characters in field names
+<!-- NEED CLARIFICATION: this should refer to a section of the index documentation that explains allowed field names -->
```
[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

File: docs/reference/query-language.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (trinity-1686a) [reply docs/reference/query-language.md]: strings are encoded as utf8, and we sort the resulting byte arrays. so `a`=>`[97]`,`A`=>`[65]`, `à`=>`[195,160]`, our order would be `A < a < à`

[the remark this answers is not in the document]
Comment (ddelemeny) [reply docs/reference/query-language.md]: OK

**Review by fmassot — APPROVED** [review]

**Pull request merged**